alf/environments/alf_gym3_wrapper.py

```diff
@@ -381,9 +381,15 @@ def _step(self, action) -> ds.TimeStep:
 
         # Override the obervation with the previous observation if that
         # particular environment has ``first=True``.
-        for i in range(self.batch_size):
-            if first[i]:
-                observation[i] = prev_observation[i]
+
+        def __override_with_prev_observation(ob_array: np.ndarray,
+                                             prev_ob_array: np.ndarray):
+            new_ob_array = ob_array
+            new_ob_array[first] = prev_ob_array[first]
+            return new_ob_array
+
+        observation = nest.map_structure(__override_with_prev_observation,
+                                         observation, prev_observation)
```

Comment:
Could just simplify this as:

``` python 
observation = nest.map_structure(lambda ob, prev_ob: ob[first] = prev_ob[first], observation, prev_observation)
```

Reply:
IMHO `map_structure` is functional, so side effect shouldn't be allowed here. Also the suggested simplification is not clear about what is returned in `nest.map_structure` call, right?

Tried in action and confirmed the idea:

1. `map_structure` does not support side effect
2. `lambda` does not support assignment (new finding, but reasonable lol)
3. Replace the `lambda` with a function that contains assignment, `map_structure` returns `None` (also reasonable).

Reply:
> IMHO `map_structure` is functional, so side effect shouldn't be allowed here. Also the suggested simplification is not clear about what is returned in `nest.map_structure` call, right?
> 
> Tried in action and confirmed the idea:
> 
> 1. `map_structure` does not support side effect
> 2. `lambda` does not support assignment (new finding, but reasonable lol)
> 3. Replace the `lambda` with a function that contains assignment, `map_structure` returns `None` (also reasonable).

You're right. Need to use a function instead of lambda. But map_structure does allow side effect, so it's OK to do:

```python
def __override(obs, prev_obs):
    obs[first] = prev_obs[first]

nest.map_structure(__override, observation, prev_observation)
```

Reply:
Sure thing ~ Updated with the side-effect version now.